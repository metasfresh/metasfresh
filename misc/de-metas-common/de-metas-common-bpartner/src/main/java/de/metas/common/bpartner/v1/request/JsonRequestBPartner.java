/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.PARENT_SYNC_ADVISE_DOC;

import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.metas.common.rest_api.common.JsonExternalId;
import de.metas.common.rest_api.v1.JsonInvoiceRule;
import de.metas.common.rest_api.v1.SyncAdvise;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@ToString
@EqualsAndHashCode
@ApiModel(description = "Note that given the respective use-case, either one of both properties might be `null`, but not both at once.")
public class JsonRequestBPartner
{
	@ApiModelProperty(position = 10,  //
			dataType = "java.lang.String", //
			value = "This translates to `C_BPartner.ExternalId`. This is deprecated and will no longer be available with v2.")
	private JsonExternalId externalId;

	@ApiModelProperty(hidden = true)
	private boolean externalIdSet;

	@ApiModelProperty(position = 20,  //
			value = "This translates to `C_BPartner.Value`.")
	private String code;

	@ApiModelProperty(hidden = true)
	private boolean codeSet;

	@ApiModelProperty(position = 25,  //
			value = "If not specified but required (e.g. because a new partner is created), then `true` is assumed.")
	private Boolean active;

	@ApiModelProperty(hidden = true)
	private boolean activeSet;

	@ApiModelProperty(position = 30,  //
			value = "This translates to `C_BPartner.Name`.\n"
					+ "If this is empty, and a BPartner with the given `name` does not yet exist, then the request will fail.")
	private String name;

	@ApiModelProperty(hidden = true)
	private boolean nameSet;

	@ApiModelProperty(position = 40,  //
			value = "This translates to `C_BPartner.Name2`.")
	private String name2;

	@ApiModelProperty(hidden = true)
	private boolean name2Set;

	@ApiModelProperty(position = 50,  //
			value = "This translates to `C_BPartner.Name3`.")
	private String name3;

	@ApiModelProperty(hidden = true)
	private boolean name3Set;

	@ApiModelProperty(position = 60, //
			value = "This translates to `C_BPartner.CompanyName`.\n"
					+ "If set, the the respective `C_BPartner` record will also have `IsCompany='Y'`")
	private String companyName;

	@ApiModelProperty(hidden = true)
	private boolean companyNameSet;

	private Boolean vendor;

	@ApiModelProperty(hidden = true)
	private boolean vendorSet;

	private Boolean customer;

	@ApiModelProperty(hidden = true)
	private boolean customerSet;

	@ApiModelProperty(position = 70, //
			value = "This translates to `C_BPartner.BPartner_Parent_ID`. It's a this bpartner's central/parent company",//
			dataType = "java.lang.Integer")
	private JsonMetasfreshId parentId;

	@ApiModelProperty(hidden = true)
	private boolean parentIdSet;

	@ApiModelProperty(position = 80, //
			value = "This translates to `C_BPartner.Phone2`. It's this bpartner's central phone number")
	private String phone;

	@ApiModelProperty(hidden = true)
	private boolean phoneSet;

	@ApiModelProperty(position = 90)
	private String language;

	@ApiModelProperty(hidden = true)
	private boolean languageSet;

	@ApiModelProperty(position = 100, //
			value = "Optional; if specified, it will be used, e.g. when an order is created for this business partner.")
	private JsonInvoiceRule invoiceRule;

	@ApiModelProperty(hidden = true)
	private boolean invoiceRuleSet;

	@ApiModelProperty(position = 110)
	private String url;

	@ApiModelProperty(hidden = true)
	private boolean urlSet;

	@ApiModelProperty(position = 120)
	private String url2;

	@ApiModelProperty(hidden = true)
	private boolean url2Set;

	@ApiModelProperty(position = 130)
	private String url3;

	@ApiModelProperty(hidden = true)
	private boolean url3Set;

	@ApiModelProperty(position = 140, //
			value = "Name of the business partner's group")
	private String group;

	@ApiModelProperty(hidden = true)
	private boolean groupSet;

	@ApiModelProperty(position = 150, //
			value = "Translates to `C_BPartner.GlobalId`")
	private String globalId;

	@ApiModelProperty(hidden = true)
	private boolean globalIdset;

	@ApiModelProperty(position = 160, //
			value = "Translates to `C_BPartner.VATaxId`")
	private String vatId;

	@ApiModelProperty(hidden = true)
	private boolean vatIdSet;

	@ApiModelProperty(position = 170, // shall be last
			value = "Sync advise about this bPartner's individual properties.\n"
			+ "IfExists is ignored on this level!\n" + PARENT_SYNC_ADVISE_DOC)
	private SyncAdvise syncAdvise;

	@ApiModelProperty(hidden = true)
	private boolean syncAdviseSet;

	public void setExternalId(final JsonExternalId externalId)
	{
		this.externalId = externalId;
		this.externalIdSet = true;
	}

	public void setCode(final String code)
	{
		this.code = code;
		this.codeSet = true;
	}

	public void setActive(final Boolean active)
	{
		this.active = active;
		this.activeSet = true;
	}

	public void setName(final String name)
	{
		this.name = name;
		this.nameSet = true;
	}

	public void setName2(final String name2)
	{
		this.name2 = name2;
		this.name2Set = true;
	}

	public void setName3(final String name3)
	{
		this.name3 = name3;
		this.name3Set = true;
	}

	public void setCompanyName(final String companyName)
	{
		this.companyName = companyName;
		this.companyNameSet = true;
	}

	public void setVendor(final Boolean vendor)
	{
		this.vendor = vendor;
		this.vendorSet = true;
	}

	public void setCustomer(final Boolean customer)
	{
		this.customer = customer;
		this.customerSet = true;
	}

	public void setParentId(final JsonMetasfreshId parentId)
	{
		this.parentId = parentId;
		this.parentIdSet = true;
	}

	public void setPhone(final String phone)
	{
		this.phone = phone;
		this.phoneSet = true;
	}

	public void setLanguage(final String language)
	{
		this.language = language;
		this.languageSet = true;
	}

	public void setInvoiceRule(final JsonInvoiceRule invoiceRule)
	{
		this.invoiceRule = invoiceRule;
		this.invoiceRuleSet = true;
	}

	public void setUrl(final String url)
	{
		this.url = url;
		this.urlSet = true;
	}

	public void setUrl2(final String url2)
	{
		this.url2 = url2;
		this.url2Set = true;
	}

	public void setUrl3(final String url3)
	{
		this.url3 = url3;
		this.url3Set = true;
	}

	public void setGroup(final String group)
	{
		this.group = group;
		this.groupSet = true;
	}

	public void setGlobalId(final String globalId)
	{
		this.globalId = globalId;
		this.globalIdset = true;
	}

	public void setSyncAdvise(final SyncAdvise syncAdvise)
	{
		this.syncAdvise = syncAdvise;
		this.syncAdviseSet = true;
	}

	public void setVatId(final String vatId)
	{
		this.vatId = vatId;
		this.vatIdSet = true;
	}
}
