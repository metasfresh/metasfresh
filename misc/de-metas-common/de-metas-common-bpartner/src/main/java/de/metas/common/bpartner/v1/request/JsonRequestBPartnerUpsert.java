/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.util.CoalesceUtil.coalesce;
import static de.metas.common.rest_api.v1.SwaggerDocConstants.*;
import java.util.List;

import javax.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableList;

import de.metas.common.rest_api.v1.SyncAdvise;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.Singular;
import lombok.Value;

@Value
@ApiModel
public class JsonRequestBPartnerUpsert
{
	@ApiModelProperty(position = 10)
	List<JsonRequestBPartnerUpsertItem> requestItems;

	@ApiModelProperty(position = 20, value = "Default sync-advise that can be overridden by individual items\n" + READ_ONLY_SYNC_ADVISE_DOC)
	SyncAdvise syncAdvise;

	@JsonCreator
	@Builder
	public JsonRequestBPartnerUpsert(
			@Singular @JsonProperty("requestItems") final List<JsonRequestBPartnerUpsertItem> requestItems,
			@Nullable @JsonProperty("syncAdvise") final SyncAdvise syncAdvise)
	{
		this.requestItems = coalesce(requestItems, ImmutableList.of());
		this.syncAdvise = coalesce(syncAdvise, SyncAdvise.READ_ONLY);
	}
}
