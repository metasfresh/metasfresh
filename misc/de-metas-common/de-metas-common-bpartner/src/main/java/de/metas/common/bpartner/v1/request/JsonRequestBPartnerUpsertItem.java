/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.BPARTNER_IDENTIFIER_DOC;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

@Value
@Builder(toBuilder = true)
@ApiModel(description = "Contains an external id and the actual bpartner to insert or update. The response will contain the given external id.")
public class JsonRequestBPartnerUpsertItem
{
	@ApiModelProperty(allowEmptyValue = false, //
			position = 10,
			value = BPARTNER_IDENTIFIER_DOC) //
	@NonNull
	String bpartnerIdentifier;

	@ApiModelProperty(allowEmptyValue = false, //
			position = 20,
			value = "The business partner to upsert. Note that its `externalId` is ignored in favor of this upsertRequest's `externalId`")
	@NonNull
	JsonRequestComposite bpartnerComposite;

	@JsonCreator
	public JsonRequestBPartnerUpsertItem(
			@NonNull @JsonProperty("bpartnerIdentifier") final String bpartnerIdentifier,
			@NonNull @JsonProperty("bpartnerComposite") final JsonRequestComposite bpartnerComposite)
	{
		this.bpartnerIdentifier = bpartnerIdentifier;
		this.bpartnerComposite = bpartnerComposite;
	}
}
