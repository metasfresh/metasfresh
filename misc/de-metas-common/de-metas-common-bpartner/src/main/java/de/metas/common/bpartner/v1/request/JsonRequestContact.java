/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.metas.common.rest_api.common.JsonExternalId;
import de.metas.common.rest_api.v1.SyncAdvise;
import io.swagger.annotations.ApiModelProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.PARENT_SYNC_ADVISE_DOC;

@Getter
@ToString
@EqualsAndHashCode
public class JsonRequestContact
{
	@ApiModelProperty(position = 10, dataType = "java.lang.String")
	private JsonExternalId externalId;

	@ApiModelProperty(hidden = true)
	private boolean externalIdSet;

	@ApiModelProperty(position = 20, dataType = "java.lang.Integer")
	private JsonMetasfreshId metasfreshBPartnerId;

	@ApiModelProperty(hidden = true)
	private boolean metasfreshBPartnerIdSet;

	@ApiModelProperty(position = 30, value = "Translated to `AD_User.Value`")
	private String code;

	@ApiModelProperty(hidden = true)
	private boolean codeSet;

	@ApiModelProperty(position = 40, value = "If not specified but required (e.g. because a new contact is created), then `true` is assumed")
	private Boolean active;

	@ApiModelProperty(hidden = true)
	private boolean activeSet;

	@ApiModelProperty(position = 50)
	private String name;

	@ApiModelProperty(hidden = true)
	private boolean nameSet;

	@ApiModelProperty(position = 60)
	private String firstName;

	@ApiModelProperty(hidden = true)
	private boolean firstNameSet;

	@ApiModelProperty(position = 70)
	private String lastName;

	@ApiModelProperty(hidden = true)
	private boolean lastNameSet;

	@ApiModelProperty(position = 80)
	private String email;

	@ApiModelProperty(hidden = true)
	private boolean emailSet;

	@ApiModelProperty(position = 90)
	private String phone;

	@ApiModelProperty(hidden = true)
	private boolean phoneSet;

	@ApiModelProperty(position = 100)
	private String fax;

	@ApiModelProperty(hidden = true)
	private boolean faxSet;

	@ApiModelProperty(position = 110)
	private String mobilePhone;

	@ApiModelProperty(hidden = true)
	private boolean mobilePhoneSet;

	@ApiModelProperty(position = 120, value = "If not specified but required (e.g. because a new contact is created), then `false` is assumed")
	private Boolean defaultContact;

	@ApiModelProperty(hidden = true)
	private boolean defaultContactSet;

	@ApiModelProperty(position = 130, //
			value = "Only one location per request may have `shipToDefault == true`.\n"
					+ "If not specified but required (e.g. because a new contact is created), then `false` is assumed.\n"
					+ "If `true`, then " //
					+ "* another possibly exiting metasfresh contact might be set to `shipToDefault = false`, even if it is not specified in this request.")
	private Boolean shipToDefault;

	@ApiModelProperty(hidden = true)
	private boolean shipToDefaultSet;

	@ApiModelProperty(position = 140, //
			value = "Only one location per request may have `billToDefault == true`.\n"
					+ "If not specified but required (e.g. because a new contact is created), then `false` is assumed.\n"
					+ "If `true`, then " //
					+ "* another possibly exiting metasfresh contact might be set to `billToDefault = false`, even if it is not specified in this request.")
	private Boolean billToDefault;

	@ApiModelProperty(hidden = true)
	private boolean billToDefaultSet;

	@ApiModelProperty(position = 150, //
			value = "If not specified but required (e.g. because a new contact is created), then `false` is assumed")
	private Boolean newsletter;

	@ApiModelProperty(hidden = true)
	private boolean newsletterSet;

	@ApiModelProperty(position = 160)
	private String description;

	@ApiModelProperty(hidden = true)
	private boolean descriptionSet;

	@ApiModelProperty(position = 170, //
			value = "If not specified but required (e.g. because a new contact is created), then `false` is assumed")
	private Boolean sales;

	@ApiModelProperty(hidden = true)
	private boolean salesSet;

	@ApiModelProperty(position = 180, //
			value = "Only one location per request may have `salesDefault == true`.\n"
					+ "If not specified but required (e.g. because a new contact is created), then `false` is assumed.\n"
					+ "If `true`, then " //
					+ "* `sales` is always be assumed to be `true` as well"
					+ "* another possibly exiting metasfresh contact might be set to `salesDefault = false`, even if it is not specified in this request.")
	private Boolean salesDefault;

	@ApiModelProperty(hidden = true)
	private boolean salesDefaultSet;

	@ApiModelProperty(position = 190, //
			value = "If not specified but required (e.g. because a new contact is created), then `false` is assumed")
	private Boolean purchase;

	@ApiModelProperty(hidden = true)
	private boolean purchaseSet;

	@ApiModelProperty(position = 200, //
			value = "Only one location per request may have `purchaseDefault == true`.\n"
					+ "If not specified but required (e.g. because a new contact is created), then `false` is assumed.\n"
					+ "If `true`, then " //
					+ "* `purchase` is always be assumed to be `true` as well"
					+ "* another possibly exiting metasfresh contact might be set to `purchaseDefault = false`, even if it is not specified in this request.")
	private Boolean purchaseDefault;

	@ApiModelProperty(hidden = true)
	private boolean purchaseDefaultSet;

	@ApiModelProperty(position = 210, //
			value = "If not specified but required (e.g. because a new contact is created), then `false` is assumed")
	private Boolean subjectMatter;

	@ApiModelProperty(hidden = true)
	private boolean subjectMatterSet;

	@ApiModelProperty(position = 220, // shall be last
			value = "Sync advise about this contact's individual properties.\n"
					+ "IfExists is ignored on this level!\n" + PARENT_SYNC_ADVISE_DOC)
	private SyncAdvise syncAdvise;

	@ApiModelProperty(hidden = true)
	private boolean syncAdviseSet;

	public void setExternalId(final JsonExternalId externalId)
	{
		this.externalId = externalId;
		this.externalIdSet = true;
	}

	public void setMetasfreshBPartnerId(final JsonMetasfreshId metasfreshBPartnerId)
	{
		this.metasfreshBPartnerId = metasfreshBPartnerId;
		this.metasfreshBPartnerIdSet = true;
	}

	public void setCode(final String code)
	{
		this.code = code;
		this.codeSet = true;
	}

	public void setActive(final Boolean active)
	{
		this.active = active;
		this.activeSet = true;
	}

	public void setName(final String name)
	{
		this.name = name;
		this.nameSet = true;
	}

	public void setFirstName(final String firstName)
	{
		this.firstName = firstName;
		this.firstNameSet = true;
	}

	public void setLastName(final String lastName)
	{
		this.lastName = lastName;
		this.lastNameSet = true;
	}

	public void setEmail(final String email)
	{
		this.email = email;
		this.emailSet = true;
	}

	public void setPhone(final String phone)
	{
		this.phone = phone;
		this.phoneSet = true;
	}

	public void setFax(final String fax)
	{
		this.fax = fax;
		this.faxSet = true;
	}

	public void setMobilePhone(final String mobilePhone)
	{
		this.mobilePhone = mobilePhone;
		this.mobilePhoneSet = true;
	}

	public void setDefaultContact(final Boolean defaultContact)
	{
		this.defaultContact = defaultContact;
		this.defaultContactSet = true;
	}

	public void setShipToDefault(final Boolean shipToDefault)
	{
		this.shipToDefault = shipToDefault;
		this.shipToDefaultSet = true;
	}

	public void setBillToDefault(final Boolean billToDefault)
	{
		this.billToDefault = billToDefault;
		this.billToDefaultSet = true;
	}

	public void setNewsletter(final Boolean newsletter)
	{
		this.newsletter = newsletter;
		this.newsletterSet = true;
	}

	public void setDescription(final String description)
	{
		this.description = description;
		this.descriptionSet = true;
	}

	public void setSales(final Boolean sales)
	{
		this.sales = sales;
		this.salesSet = true;
	}

	public void setSalesDefault(final Boolean salesDefault)
	{
		this.salesDefault = salesDefault;
		this.salesDefaultSet = true;
	}

	public void setPurchase(final Boolean purchase)
	{
		this.purchase = purchase;
		this.purchaseSet = true;
	}

	public void setPurchaseDefault(final Boolean purchaseDefault)
	{
		this.purchaseDefault = purchaseDefault;
		this.purchaseDefaultSet = true;
	}

	public void setSubjectMatter(final Boolean subjectMatter)
	{
		this.subjectMatter = subjectMatter;
		this.subjectMatterSet = true;
	}

	public void setSyncAdvise(final SyncAdvise syncAdvise)
	{
		this.syncAdvise = syncAdvise;
		this.syncAdviseSet = true;
	}
}
