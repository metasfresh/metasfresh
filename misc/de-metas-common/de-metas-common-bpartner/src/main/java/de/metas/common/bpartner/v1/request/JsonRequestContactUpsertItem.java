/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.*;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

@Value
@Builder(toBuilder = true)
@ApiModel(description = "Contains an external id and the actual bpartner to insert or update. The response will contain the given external id.")
public class JsonRequestContactUpsertItem
{
	@ApiModelProperty(allowEmptyValue = false, position = 10, //
			value = CONTACT_IDENTIFIER_DOC + "\n"//
					+ "If the identifier is an `<AD_User_ID>`, then it is assumed that the resource exists in metasfresh.\n"
					+ "If a new contact is created and the actual contact has no different identifier, then this identifier is stored within the newly created contact."
					+ "This is deprecated and will no longer be available with v2.") //
	@NonNull
	final String contactIdentifier;

	@ApiModelProperty(allowEmptyValue = false, //
			position = 20, value = "The contact to upsert. Note that its `externalId` is ignored in favor of this upsertRequest's `externalId`")
	@NonNull
	JsonRequestContact contact;

	@JsonCreator
	public JsonRequestContactUpsertItem(
			@NonNull @JsonProperty("contactIdentifier") String contactIdentifier,
			@NonNull @JsonProperty("contact") JsonRequestContact contact)
	{
		this.contactIdentifier = contactIdentifier;
		this.contact = contact;
	}
}
