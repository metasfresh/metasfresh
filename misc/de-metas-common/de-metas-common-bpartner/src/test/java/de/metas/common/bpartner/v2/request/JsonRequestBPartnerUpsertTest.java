/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.request;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.metas.common.bpartner.v1.request.JsonRequestBPartnerUpsert;
import de.metas.common.bpartner.v1.request.JsonRequestContact;
import lombok.NonNull;
import org.apache.commons.io.IOUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.io.IOException;
import java.io.InputStream;

import static org.assertj.core.api.Assertions.*;

@ExtendWith(SnapshotExtension.class)
public class JsonRequestBPartnerUpsertTest
{
	private final ObjectMapper mapper = new ObjectMapper();
	private Expect expect;

	@Test
	public void deserialize_1() throws IOException
	{
		final JsonRequestBPartnerUpsert result = deserialize("/de/metas/common/bpartner/v1/request/JsonBPartnerUpsertRequest_1.json");

		final JsonRequestContact contact = result.getRequestItems().get(0).getBpartnerComposite().getContactsNotNull().getRequestItems().get(0).getContact();

		// test one particular case were we explicitly have "fax" : null in the JSON
		assertThat(contact.getFax()).isNull();
		assertThat(contact.isFaxSet()).isTrue();

		expect.serializer("orderedJson").toMatchSnapshot(result);
	}

	private JsonRequestBPartnerUpsert deserialize(@NonNull final String jsonResourceName) throws IOException
	{
		final JsonRequestBPartnerUpsert request = readInstanceFromFile(jsonResourceName);

		final String string = mapper.writeValueAsString(request);
		final JsonRequestBPartnerUpsert result = mapper.readValue(string, JsonRequestBPartnerUpsert.class);

		assertThat(result).isEqualTo(request);
		return result;
	}

	private JsonRequestBPartnerUpsert readInstanceFromFile(@NonNull final String jsonResourceName) throws IOException
	{

		final InputStream stream = JsonRequestBPartnerUpsertTest.class.getResourceAsStream(jsonResourceName);
		
		final String string = IOUtils.toString(stream, "UTF-8");
		assertThat(string).isNotNull(); // guard
		
		final JsonRequestBPartnerUpsert request = mapper.readValue(string,JsonRequestBPartnerUpsert.class);

		return request;
	}
}
