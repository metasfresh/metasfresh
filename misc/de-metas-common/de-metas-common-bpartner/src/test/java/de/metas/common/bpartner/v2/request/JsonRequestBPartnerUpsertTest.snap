de.metas.common.bpartner.v2.request.JsonRequestBPartnerUpsertTest.deserialize_1=[
  {
    "requestItems": [
      {
        "bpartnerComposite": {
          "bpartner": {
            "activeSet": false,
            "code": "code",
            "codeSet": true,
            "companyName": "bPartner.companyName",
            "companyNameSet": true,
            "customerSet": false,
            "externalId": "fghij_2",
            "externalIdSet": true,
            "globalIdset": false,
            "group": "bpGroupRecord.name",
            "groupSet": true,
            "invoiceRuleSet": false,
            "language": "bPartner.de_CH",
            "languageSet": true,
            "name": "bPartner.name",
            "name2Set": false,
            "name3Set": false,
            "nameSet": true,
            "parentId": 1,
            "parentIdSet": true,
            "phone": "bPartner.phone",
            "phoneSet": true,
            "syncAdviseSet": false,
            "url": "bPartner.url",
            "url2Set": false,
            "url3Set": false,
            "urlSet": true,
            "vatId": "bPartner.vatId",
            "vatIdSet": true,
            "vendorSet": false
          },
          "contacts": {
            "requestItems": [
              {
                "contact": {
                  "activeSet": false,
                  "billToDefaultSet": false,
                  "codeSet": false,
                  "defaultContactSet": false,
                  "descriptionSet": false,
                  "email": "c1_email@email.net",
                  "emailSet": true,
                  "externalId": "c1_externalId",
                  "externalIdSet": true,
                  "faxSet": true,
                  "firstNameSet": false,
                  "lastNameSet": false,
                  "metasfreshBPartnerIdSet": false,
                  "mobilePhoneSet": false,
                  "name": "c1_name",
                  "nameSet": true,
                  "newsletterSet": false,
                  "phone": "c1_phone",
                  "phoneSet": true,
                  "purchaseDefaultSet": false,
                  "purchaseSet": false,
                  "salesDefaultSet": false,
                  "salesSet": false,
                  "shipToDefaultSet": false,
                  "subjectMatterSet": false,
                  "syncAdviseSet": false
                },
                "contactIdentifier": "ext-c1_externalId"
              },
              {
                "contact": {
                  "activeSet": false,
                  "billToDefaultSet": false,
                  "codeSet": false,
                  "defaultContactSet": false,
                  "descriptionSet": false,
                  "email": "c2_email@email.net",
                  "emailSet": true,
                  "externalId": "c2_externalId",
                  "externalIdSet": true,
                  "faxSet": false,
                  "firstNameSet": false,
                  "lastNameSet": false,
                  "metasfreshBPartnerIdSet": false,
                  "mobilePhoneSet": false,
                  "name": "c2_name",
                  "nameSet": true,
                  "newsletterSet": false,
                  "phone": "c2_phone",
                  "phoneSet": true,
                  "purchaseDefaultSet": false,
                  "purchaseSet": false,
                  "salesDefaultSet": false,
                  "salesSet": false,
                  "shipToDefaultSet": false,
                  "subjectMatterSet": false,
                  "syncAdviseSet": false
                },
                "contactIdentifier": "ext-c2_externalId"
              }
            ],
            "syncAdvise": {
              "ifExists": "UPDATE_MERGE",
              "ifNotExists": "CREATE"
            }
          },
          "locations": {
            "requestItems": [
              {
                "location": {
                  "activeSet": false,
                  "address1": "l1_address1",
                  "address1Set": true,
                  "address2": "l1_address2",
                  "address2Set": true,
                  "address3Set": false,
                  "address4Set": false,
                  "billToDefaultSet": false,
                  "billToSet": false,
                  "bpartnerNameSet": false,
                  "city": "l1_city",
                  "citySet": true,
                  "countryCode": "CH",
                  "countryCodeSet": true,
                  "district": "l1_district",
                  "districtSet": true,
                  "externalId": "l1_externalId",
                  "externalIdSet": true,
                  "gln": "l1_gln",
                  "glnSet": true,
                  "nameSet": false,
                  "poBox": "l1_poBox",
                  "poBoxSet": true,
                  "postal": "l1_postal",
                  "postalSet": true,
                  "region": "l1_region",
                  "regionSet": true,
                  "shipToDefaultSet": false,
                  "shipToSet": false,
                  "syncAdviseSet": false
                },
                "locationIdentifier": "ext-l1_externalId"
              },
              {
                "location": {
                  "activeSet": false,
                  "address1": "l2_address1",
                  "address1Set": true,
                  "address2": "l2_address2",
                  "address2Set": true,
                  "address3Set": false,
                  "address4Set": false,
                  "billToDefaultSet": false,
                  "billToSet": false,
                  "bpartnerNameSet": false,
                  "city": "l2_city",
                  "citySet": true,
                  "countryCode": "DE",
                  "countryCodeSet": true,
                  "district": "l2_district",
                  "districtSet": true,
                  "externalId": "l2_externalId",
                  "externalIdSet": true,
                  "gln": "l2_gln",
                  "glnSet": true,
                  "nameSet": false,
                  "poBox": "l2_poBox",
                  "poBoxSet": true,
                  "postal": "l2_postal",
                  "postalSet": true,
                  "region": "l2_region",
                  "regionSet": true,
                  "shipToDefaultSet": false,
                  "shipToSet": false,
                  "syncAdviseSet": false
                },
                "locationIdentifier": "ext-l2_externalId"
              }
            ],
            "syncAdvise": {
              "ifExists": "UPDATE_MERGE",
              "ifNotExists": "CREATE"
            }
          }
        },
        "bpartnerIdentifier": "ext-fghij_2"
      }
    ],
    "syncAdvise": {
      "ifExists": "UPDATE_MERGE",
      "ifNotExists": "CREATE"
    }
  }
]